[Merge] 专业管理相关操作

fanfan187 opened the pull request:
---

## 变更内容

- **新增/修改/删除**：
    - 添加专业相关数据传输对象 `MajorDTO`、`MajorAcademicDTO`、`MajorStudentDTO`
    - 实现专业管理的业务逻辑 `MajorLogic`
    - 添加专业管理的控制器 `MajorController`
    - 在 `DepartmentMapper` 中增加根据部门名称获取部门 UUID 的方法
    - 新增 `MajorVO` 类作为专业视图对象

fanfan187 committed:
feat(major): 新增专业管理功能

- 添加专业相关数据传输对象 `MajorDTO`、`MajorAcademicDTO`、`MajorStudentDTO`
- 实现专业管理的业务逻辑 `MajorLogic`
- 添加专业管理的控制器 `MajorController`
- 在 `DepartmentMapper` 中增加根据部门名称获取部门 UUID 的方法

fanfan187 committed:
refactor(major): 重构专业相关接口
- 新增 `MajorVO` 类作为专业视图对象
- 修改 addMajor 和 updateMajor 方法，使用 `MajorVO` 作为参数
- 在方法内部将 `MajorVO` 转换为 `MajorDTO`
- 优化代码结构，提高可维护性

xiaolfeng commented on src/main/java/com/frontleaves/scheduling/controllers/MajorController.java:
```diff
@@ -0,0 +1,199 @@
+            @PathVariable("major_uuid") String majorUuid,
+            @RequestBody MajorVO majorVO
+    ) throws NotFoundException {
+        log.info("===> 进入 updateMajor 方法, majorUuid: {}", majorUuid);
```
对于日志使用 info 时候意思是可以告诉客户这个日志内容，如果只是开发人员测试用的应该是用 debug 或者 trace（这个只是提醒，可以选择不修改）

xiaolfeng commented on src/main/java/com/frontleaves/scheduling/controllers/MajorController.java:
```diff
@@ -0,0 +1,199 @@
+        // VO -> DTO
+        MajorDTO majorDTO = new MajorDTO();
+        BeanUtils.copyProperties(majorVO, majorDTO);
+        MajorDTO updated = majorService.updateMajor(majorUuid, majorDTO);
```
majorUuid 没有正则判断，如果输入错的值会导致最终查询出现问题

xiaolfeng commented on src/main/java/com/frontleaves/scheduling/controllers/MajorController.java:
```diff
@@ -0,0 +1,199 @@
+            @PathVariable("major_uuid")
+            String majorUuid
+    ) throws NotFoundException {
+        majorService.deleteMajor(majorUuid);
```
正则判断

xiaolfeng commented on src/main/java/com/frontleaves/scheduling/controllers/MajorController.java:
```diff
@@ -0,0 +1,199 @@
+    @GetMapping("/{major_uuid}")
+    public ResponseEntity<BaseResponse<MajorDTO>> getMajor(
+            @PathVariable("major_uuid") String majorUuid
+    ) throws NotFoundException {
```
正则判断

xiaolfeng commented on src/main/java/com/frontleaves/scheduling/logic/MajorLogic.java:
```diff
@@ -0,0 +1,249 @@
+public class MajorLogic implements MajorService {
+    private final MajorDAO majorDAO;
+    private final DepartmentMapper departmentMapper;
+    private static final String MAJOR_NOT_FOUND = "专业不存在";
```
写在 StringConstant 里

xiaolfeng commented on src/main/java/com/frontleaves/scheduling/logic/MajorLogic.java:
```diff
@@ -0,0 +1,249 @@
+        MajorDO majorDO = majorDAO.getById(majorUuid);
+        // 验证专业存在
+        if (majorDO == null) {
+            throw new NotFoundException(MAJOR_NOT_FOUND);
```
业务抛出请使用 BusinessException

xiaolfeng commented on src/main/java/com/frontleaves/scheduling/models/vo/MajorVO.java:
```diff
@@ -0,0 +1,57 @@
+@NoArgsConstructor
+@AllArgsConstructor
+@Accessors(chain = true)
+public class MajorVO {
```
没有进行数据校验，若输入数据超出数据库限制，导致报错（没有进行错误数据校验）

xiaolfeng commented on src/main/java/com/frontleaves/scheduling/logic/MajorLogic.java:
```diff
@@ -0,0 +1,249 @@
+
+        // 把 DTO 里的变动字段覆盖上去
+        BeanUtils.copyProperties(majorDTO, majorDO, "majorUuid", "createAt", "updatedAt");
+        majorDO.setMajorUuid(majorUuid);
```
主键禁止修改！（一般情况下只用做查询，不允许 uuid 类型主键修改）

xiaolfeng commented on src/main/java/com/frontleaves/scheduling/logic/MajorLogic.java:
```diff
@@ -0,0 +1,249 @@
+        // 把 DTO 里的变动字段覆盖上去
+        BeanUtils.copyProperties(majorDTO, majorDO, "majorUuid", "createAt", "updatedAt");
+        majorDO.setMajorUuid(majorUuid);
+        majorDAO.updateById(majorDO);
```
在这里如果 uuid 修改导致更新失败（因为修改后的 uuid ，在数据库中可能不存在）

xiaolfeng commented on src/main/java/com/frontleaves/scheduling/logic/MajorLogic.java:
```diff
@@ -0,0 +1,249 @@
+     */
+    @Override
+    public MajorDTO updateMajor(String majorUuid, MajorDTO majorDTO) throws NotFoundException {
+        MajorDO majorDO = majorDAO.getById(majorUuid);
```
数据库操作应该由 DAO 操作，应该定义一个方法 `majorDAO.getMajorByUuid(String majorUuid)` 并在这个方法内首先从缓存获取数据，若查询不到则从数据库获取数据

xiaolfeng commented on src/main/java/com/frontleaves/scheduling/logic/MajorLogic.java:
```diff
@@ -0,0 +1,249 @@
+    @Override
+    public void deleteMajor(String majorUuid) throws NotFoundException {
+        // 检查专业是否存在
+        MajorDO majorDO = majorDAO.getById(majorUuid);
```
同理

xiaolfeng commented on src/main/java/com/frontleaves/scheduling/logic/MajorLogic.java:
```diff
@@ -0,0 +1,249 @@
+        // 检查专业是否存在
+        MajorDO majorDO = majorDAO.getById(majorUuid);
+        if (majorDO == null) {
+            throw new NotFoundException(MAJOR_NOT_FOUND);
```
同理，以下部分均同理

xiaolfeng commented on src/main/java/com/frontleaves/scheduling/logic/MajorLogic.java:
```diff
@@ -0,0 +1,249 @@
+        }
+
+        // 检查专业是否被系统引用
+        if (majorDO.getMajorStatus() == 1) {
```
是否被引用应当由数据库外键约束或查询被约束数据库检查是否存在数据。状态是提供用户显示这个专业是否可下次被选择（禁用以后不可以被其他关联，但是已关联不会取消关联）

xiaolfeng commented on src/main/java/com/frontleaves/scheduling/logic/MajorLogic.java:
```diff
@@ -0,0 +1,249 @@
+        }
+
+        // 进行二次确认
+        log.warn("即将删除专业[{}],请确认是否继续", majorUuid);
```
二次确认这个应该是 AI 写多了的，二次确认是前端进行操作的，数据库直接删除就可以了

xiaolfeng commented on src/main/java/com/frontleaves/scheduling/logic/MajorLogic.java:
```diff
@@ -0,0 +1,249 @@
+     */
+    @Override
+    public PageDTO<MajorDTO> listMajorsForAdmin(int page, int size, Boolean isDesc, String department, String name) {
+        return listMajors(page, size, isDesc, department, name, majorDO -> {
```
用 lambda 这个来做 map 操作想法很好，不过在数据库操作中为了保证数据操作的原子化，DAO 输出的数据有可能被其他 Logic 使用，需要保证数据原子化。并且获取的数据应该是属于原始数据（也就是说，输出应该是用 mybatisPlus 的 Page，泛型是用 xxxDO 进行输出，格式转换应该由 Logic 进行，DAO 只处理对数据库的操作，一切逻辑操作交给 Logic）

xiaolfeng commented on src/main/java/com/frontleaves/scheduling/mappers/DepartmentMapper.java:
```diff
@@ -3,11 +3,26 @@
+        FROM cs_department
+        WHERE department_name LIKE CONCAT('%', #{departmentName}, '%')
+""")
+    List<String> getDepartmentUuidByName(String departmentName);
```
可以直接使用 MybatisPlus

xiaolfeng reviewed (changes requested):
有内容需要进行修改，注意系统各模块之间保证相对独立性（合理解耦）

DAO 需要保证数据库操作原子化，并且使用缓存机制。
Logic 专门处理数据逻辑
Controller 接收数据，对数据进行校验以及引导如何进行 Logic 操作

fanfan187 committed:
feat(major): 增加专业信息缓存和引用检查功能

- 在 `MajorDAO` 中添加了基于 Redis 的专业信息缓存功能
- 实现了专业信息的引用检查方法
- 优化了专业列表查询逻辑，移除了冗余的 SQL 查询
- 更新了 `MajorService` 接口，简化了方法签名
- 在 `MajorVO` 中添加了字段验证注解，提高了数据输入的准确性

fanfan187 committed:
Merge remote-tracking branch 'origin/develop' into feature/major-operate

# Conflicts:
#	src/main/java/com/frontleaves/scheduling/constants/StringConstant.java

xiaolfeng commented on src/main/java/com/frontleaves/scheduling/controllers/MajorController.java:
```diff
@@ -0,0 +1,235 @@
+            String majorUuid
+    ) {
+        // 业务流程，使用 info 级别
+        log.info("删除专业: {}", majorUuid);
```
1

fanfan187 replied:
这个我看到了，我想的是除去前端直接看到外，这个不是也属于业务层面的吗🤔

xiaolfeng replied:
这个在 AOP 切面已经捕获数据传输了，在 DAO 那边会显示输出

xiaolfeng replied:
所以也属于开发者测试数据内容显示的部分

xiaolfeng commented on src/main/java/com/frontleaves/scheduling/mappers/DepartmentMapper.java:
```diff
@@ -1,13 +1,28 @@
+     *
+     * @param departmentName 部门名称
+     */
+    default List<String> getDepartmentUuidByName(String departmentName) {
```
解决 develop 冲突后 DAO 有这个方法

xiaolfeng commented on src/main/java/com/frontleaves/scheduling/daos/MajorDAO.java:
```diff
@@ -48,4 +64,71 @@
+        LambdaQueryWrapper<MajorDO> queryWrapper = new LambdaQueryWrapper<>();
+
+        if (CharSequenceUtil.isNotBlank(department)) {
+            List<String> departmentUuids = departmentMapper.getDepartmentUuidByName(department);
```
MybatisPlus 方法应该是用 DAO，在 Mapper 是用属于侵入

fanfan187 committed:
Merge remote-tracking branch 'origin/develop' into feature/major-operate

# Conflicts:
#	src/main/java/com/frontleaves/scheduling/constants/StringConstant.java
#	src/main/java/com/frontleaves/scheduling/models/entity/DepartmentDO.java

fanfan187 committed:
refactor(major): 重构专业相关功能并添加单元测试

- 移除 `DepartmentMapper` 中的 getDepartmentUuidByName 方法，改由 `DepartmentDAO` 提供
- 重构 `MajorDAO` 中的 isReferenced 方法，改为 isReferenced 和 deleteMajor两个独立方法
- 优化 `MajorController` 中的 UUID 正则匹配逻辑
- 新增 `MajorTest` 类，实现专业相关功能的单元测试
- 更新 `StringConstant` 中的 MAJOR_UUID 常量，添加冒号后缀

xiaolfeng committed:
Merge branch 'develop' into feature/major-operate
